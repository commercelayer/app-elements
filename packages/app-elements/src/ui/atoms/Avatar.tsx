import cn from "classnames"
import { type JSX, useState } from "react"
import { presets } from "./Avatar.utils"

type SrcPreset = keyof typeof presets
type SrcUrl = `https://${string}` | `data:image/${string}`

export interface AvatarProps {
  /**
   * Image URL
   */
  src: SrcPreset | SrcUrl
  /**
   * Alt text
   */
  alt: string
  /**
   * Specify `none` to remove border
   */
  border?: "none"
  /**
   * Image shape
   * @default "rounded"
   */
  shape?: "rounded" | "circle"
  /**
   * Image size
   * (x-small: 32px, small: 48px, normal: 58px, large: 72px)
   * @default "normal"
   */
  size?: "x-small" | "small" | "normal" | "large"
  /**
   * Image class
   */
  className?: string
}

/**
 * This component renders as `<img>` using different shapes and sizes. It is mostly used to show an SKU image or a 3rd-party icon.
 */
export function Avatar({
  src,
  alt,
  border,
  shape = "rounded",
  size = "normal",
  className,
  ...rest
}: AvatarProps): JSX.Element {
  const [hasOnLoadError, setHasOnLoadError] = useState<boolean>(false)
  const hasError =
    hasOnLoadError || (!srcIsValidUrl(src) && !srcIsValidPreset(src))

  return (
    <img
      {...rest}
      onError={() => {
        setHasOnLoadError(true)
      }}
      src={
        srcIsValidPreset(src)
          ? presets[src]
          : srcIsValidUrl(src) && !hasOnLoadError
            ? src
            : placeholderSvg
      }
      alt={alt}
      className={cn(
        "border object-contain object-center",
        {
          // size
          "min-w-[72px] min-h-[72px] w-[72px] h-[72px]": size === "large",
          "min-w-[58px] min-h-[58px] w-[58px] h-[58px]": size === "normal",
          "min-w-[42px] min-h-[42px] w-[42px] h-[42px]": size === "small",
          "min-w-8 min-h-8 w-8 h-8": size === "x-small",
          // shape
          rounded: shape === "rounded",
          "rounded-full": shape === "circle",
          // border
          "border-gray-100": border == null,
          "border-transparent":
            border === "none" || (srcIsValidPreset(src) && src !== "gift_card"),
          // placeholder
          "p-1": hasError && (size === "normal" || size === "large"),
          "p-0.5": hasError && size !== "normal",
        },
        className,
      )}
    />
  )
}

function srcIsValidPreset(src: AvatarProps["src"]): src is SrcPreset {
  return Object.keys(presets).includes(src)
}

function srcIsValidUrl(
  src: AvatarProps["src"] | undefined | null,
): src is SrcUrl {
  return (
    src != null && (src.startsWith("https://") || src.startsWith("data:image/"))
  )
}

const placeholderSvg =
  "https://data.commercelayer.app/assets/images/icons/items/placeholder.svg"

Avatar.displayName = "Avatar"
