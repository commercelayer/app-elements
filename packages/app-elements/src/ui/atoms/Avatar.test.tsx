import { render, type RenderResult } from '@testing-library/react'
import { Avatar, type AvatarProps } from './Avatar'

interface SetupProps extends AvatarProps {
  id: string
}

type SetupResult = RenderResult & {
  element: HTMLElement
}

const setup = ({ id, ...props }: SetupProps): SetupResult => {
  const utils = render(<Avatar data-test-id={id} {...props} />)
  const element = utils.getByTestId(id)
  return {
    element,
    ...utils
  }
}

describe('Avatar', () => {
  test('Should be rendered', () => {
    const { element } = setup({
      id: 'avatar',
      src: 'https://i2.wp.com/ui-avatars.com/api/Commerce+Layer/160/FF656B/FFFFFF/2/0.33/true/true/true?ssl=1',
      alt: 'Commerce Layer logo'
    })
    expect(element).toBeVisible()
    expect(element).toMatchSnapshot()
    expect(element.getAttribute('alt')).toBe('Commerce Layer logo')
  })
})
