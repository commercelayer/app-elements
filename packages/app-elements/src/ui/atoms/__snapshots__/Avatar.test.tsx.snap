// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Avatar > Should be rendered with a placeholder image when src is not defined 1`] = `
<img
  alt="Undefined source"
  class="border object-contain object-center min-w-[58px] min-h-[58px] w-[58px] h-[58px] rounded border-gray-100 p-1"
  data-testid="avatar"
  src="https://data.commercelayer.app/assets/images/icons/items/placeholder.svg"
/>
`;

exports[`Avatar > Should be rendered with src pointing to a preset 1`] = `
<img
  alt="Stripe"
  class="border object-contain object-center min-w-[58px] min-h-[58px] w-[58px] h-[58px] rounded border-gray-100 border-transparent"
  data-testid="avatar"
  src="https://data.commercelayer.app/assets/images/icons/payment-gateways/stripe.svg"
/>
`;

exports[`Avatar > Should be rendered with src pointing to an https url 1`] = `
<img
  alt="Commerce Layer logo"
  class="border object-contain object-center min-w-[58px] min-h-[58px] w-[58px] h-[58px] rounded border-gray-100"
  data-testid="avatar"
  src="https://i2.wp.com/ui-avatars.com/api/Commerce+Layer/160/FF656B/FFFFFF/2/0.33/true/true/true?ssl=1"
/>
`;
